llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -5534,14 +5553,23 @@ InstructionCost LoopVectorizationCostModel::computePredInstDiscount(
 InstructionCost LoopVectorizationCostModel::expectedCost(ElementCount VF) {
   InstructionCost Cost;
 
+  // If the vector loop gets executed exactly once with the given VF, ignore the
+  // costs of comparison and induction instructions, as they'll get simplified
+  // away.
+  SmallPtrSet<Instruction *, 2> ValuesToIgnoreForVF;
+  auto TC = PSE.getSE()->getSmallConstantTripCount(TheLoop);
+  if (VF.isFixed() && TC == VF.getFixedValue())
```

Comment:
I think this should check we're not tail-folding here, i.e.

`if (VF.isFixed() && TC == VF.getFixedValue() && !foldTailByMasking())`